### PR TITLE
Support get contributor from non-sxs branch

### PR DESCRIPTION
**Pull request by OsmondJiang**

### CODE REVIEW

File: src/docfx/lib/git/GitCommitProvider.cs [hunk 1]
```diff
@@ -221,15 +224,36 @@ public void Dispose()
...
+                {
+                    if (GitBranchLookup(out var refBranch, _repo, pLocaleBranchName, 1 /*locale branch*/) == 0 ||
+                        GitBranchLookup(out refBranch, _repo, pRemoteBranchName, 2 /*remote branch*/) == 0)
+                    {
+                        var commit = GitReferenceTarget(refBranch);
+                        GitRevwalkPush(walk, commit);
+                    }
+                    else
+                    {
+                        throw new ArgumentException($"{branchName} can't not be resolved");
```
Comment (yufeih) [inline src/docfx/lib/git/GitCommitProvider.cs]: we should have an user error if the loc repo or branch does not exists

[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/git/GitCommitProvider.cs]: ok, but why below Errors.GitShadowClone doesn't release walk before throw?

---
In reply to: [229592511](https://github.com/dotnet/docfx/pull/3546#discussion_r229592511) [](ancestors = 229592511)

[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/git/GitCommitProvider.cs]: and also the other invalidOperationException()?

---
In reply to: [229653306](https://github.com/dotnet/docfx/pull/3546#discussion_r229653306) [](ancestors = 229653306,229592511)

File: src/docfx/lib/git/GitCommitProvider.cs [hunk 2]
```diff
@@ -223,15 +226,36 @@ public void Dispose()
...
+            {
+                GitRevwalkPushHead(walk);
+            }
+            else
+            {
+                fixed (byte* pLocaleBranchName = ToUtf8Native($"{branchName}"))
+                fixed (byte* pRemoteBranchName = ToUtf8Native($"origin/{branchName}"))
+                {
+                    if (GitBranchLookup(out var refBranch, _repo, pLocaleBranchName, 1 /*locale branch*/) == 0 ||
+                        GitBranchLookup(out refBranch, _repo, pRemoteBranchName, 2 /*remote branch*/) == 0)
```
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/git/GitCommitProvider.cs]: if the local branch has the new commit, we should use the local branch firstly,like testing

---
In reply to: [229591569](https://github.com/dotnet/docfx/pull/3546#discussion_r229591569) [](ancestors = 229591569)

**Comment by OsmondJiang** [comment]
#3547

File: src/docfx/build/metadata/ContributionProvider.cs [hunk 3]
```diff
@@ -249,14 +249,15 @@ private async Task LoadCommits(Docset docset)
                 {
                     var repo = group.Key;
                     var repoPath = repo.Path;
+                    var repoBranch = bilingual && repo.Branch.EndsWith("-sxs") ? repo.Branch.Substring(0, repo.Branch.Length - 4) : null;
```
Comment (live1206) [inline src/docfx/build/metadata/ContributionProvider.cs]: when bilingual is true, and input repo branch not ending with "-sxs", should we log some warning to inform user?

File: src/docfx/lib/git/GitCommitProvider.cs [hunk 4]
```diff
@@ -61,13 +61,16 @@ public static async Task<GitCommitProvider> Create(string repoPath, string cache
             return new GitCommitProvider(repoPath, cacheFilePath, await LoadCommitCache(cacheFilePath));
         }
 
-        public List<GitCommit> GetCommitHistory(string file)
+        public List<GitCommit> GetCommitHistory(string file, string branch = null)
```
Comment (live1206) [inline src/docfx/lib/git/GitCommitProvider.cs]: default branch as `string.Empty`?

File: src/docfx/build/metadata/ContributionProvider.cs (hunk 3 above)
Comment (OsmondJiang) [reply src/docfx/build/metadata/ContributionProvider.cs]: u are right, but we should not log here, we should log warnings at the begining of building, let's check it in another PR

---
In reply to: [229918463](https://github.com/dotnet/docfx/pull/3546#discussion_r229918463) [](ancestors = 229918463)

File: src/docfx/lib/git/GitCommitProvider.cs (hunk 4 above)
Comment (OsmondJiang) [reply src/docfx/lib/git/GitCommitProvider.cs]: I believe both are OK, even the default branch is string.empty, user still gets chance to pass branch 'null'

---
In reply to: [229918800](https://github.com/dotnet/docfx/pull/3546#discussion_r229918800) [](ancestors = 229918800)

File: src/docfx/restore/RestoreWorkTree.cs [hunk 5]
```diff
@@ -75,7 +75,9 @@ async Task AddWorkTrees()
                     var workTreePath = GetRestoreWorkTreeDir(restoreDir, workTreeHead);
                     if (existingWorkTrees.TryAdd(workTreePath, 0))
                     {
-                        await GitUtility.AddWorkTree(restorePath, workTreeHead, workTreePath);
+                        // use branch name instead of commit hash
+                        // https://git-scm.com/docs/git-worktree#_commands
+                        await GitUtility.AddWorkTree(restorePath, rev, workTreePath);
```
Comment (yufeih) [inline src/docfx/restore/RestoreWorkTree.cs]: There is a tiny problem with this approach if this `branch` is shared globally and accessed at the same time by different builds.

Comment (OsmondJiang) [reply src/docfx/restore/RestoreWorkTree.cs]: ok

Comment (OsmondJiang) [reply src/docfx/restore/RestoreWorkTree.cs]: although we use the branch, but the worktree path is different for different build.

**Pull request merged**